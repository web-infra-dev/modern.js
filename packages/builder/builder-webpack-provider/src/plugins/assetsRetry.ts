import { BuilderPlugin } from '../types';
import { isHtmlDisabled } from './html';

export function builderPluginAssetsRetry(): BuilderPlugin {
  return {
    name: 'builder-plugin-assets-retry',
    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getNormalizedConfig();

        if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
          return;
        }

        const { AssetsRetryPlugin } = await import(
          '../webpackPlugins/AssetsRetryPlugin'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY)
          .use(AssetsRetryPlugin, [config.output.assetsRetry]);
      });
    },
  };
}
