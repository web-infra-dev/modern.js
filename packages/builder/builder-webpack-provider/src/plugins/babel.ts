import {
  getBabelConfig,
  createBabelChain,
  type BabelOptions,
} from '@modern-js/babel-preset-app';
import {
  JS_REGEX,
  TS_REGEX,
  mergeRegex,
  createVirtualModule,
  getBrowserslistWithDefault,
  type BuilderContext,
} from '@modern-js/builder-shared';

import type { WebpackChain, BuilderPlugin, NormalizedConfig } from '../types';

const enableCoreJsEntry = (config: NormalizedConfig, isServer: boolean) =>
  config.output.polyfill === 'entry' && !isServer;

export const getUseBuiltIns = (config: NormalizedConfig) => {
  const { polyfill } = config.output;
  if (polyfill === 'ua' || polyfill === 'off') {
    return false;
  }
  return polyfill;
};

export function applyScriptCondition({
  rule,
  config,
  context,
  includes,
  excludes,
}: {
  rule: WebpackChain.Rule;
  config: NormalizedConfig;
  context: BuilderContext;
  includes: (string | RegExp)[];
  excludes: (string | RegExp)[];
}) {
  // compile all folders in app directory, exclude node_modules
  rule.include.add({
    and: [context.rootPath, { not: /node_modules/ }],
  });

  [...includes, ...(config.source.include || [])].forEach(condition => {
    rule.include.add(condition);
  });

  [...excludes, ...(config.source.exclude || [])].forEach(condition => {
    rule.exclude.add(condition);
  });
}

export const builderPluginBabel = (): BuilderPlugin => ({
  name: 'builder-plugin-babel',
  setup(api) {
    api.modifyWebpackChain(
      async (
        chain,
        { CHAIN_ID, getCompiledPath, target, isProd, isServer },
      ) => {
        const { applyOptionsChain, isUseSSRBundle } = await import(
          '@modern-js/utils'
        );

        const config = api.getNormalizedConfig();
        const browserslist = await getBrowserslistWithDefault(
          api.context.rootPath,
          config,
          target,
        );

        const getBabelOptions = (
          appDirectory: string,
          config: NormalizedConfig,
        ) => {
          // 1. Get styled-components options
          const styledComponentsOptions = applyOptionsChain(
            {
              // "pure" is used to improve dead code elimination in production.
              // we don't need to enable it in development because it will slow down the build process.
              pure: isProd,
              displayName: true,
              ssr: isUseSSRBundle(config),
              transpileTemplateLiterals: true,
            },
            config.tools.styledComponents,
          );

          // 2. Create babel util function about include/exclude
          const includes: Array<string | RegExp> = [];
          const excludes: Array<string | RegExp> = [];

          const babelUtils = {
            addIncludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                includes.push(...items);
              } else {
                includes.push(items);
              }
            },
            addExcludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                excludes.push(...items);
              } else {
                excludes.push(items);
              }
            },
          };

          // 3. Compute final babel config by @modern-js/babel-preset-app
          const babelOptions: BabelOptions = {
            babelrc: false,
            configFile: false,
            compact: isProd,
            ...getBabelConfig({
              target: isServer ? 'server' : 'client',
              appDirectory,
              useLegacyDecorators: !config.output.enableLatestDecorators,
              useBuiltIns: isServer ? false : getUseBuiltIns(config),
              chain: createBabelChain(),
              styledComponents: styledComponentsOptions,
              userBabelConfig: config.tools.babel,
              userBabelConfigUtils: babelUtils,
              overrideBrowserslist: browserslist,
            }),
          };

          if (config.output.charset === 'utf8') {
            babelOptions.generatorOpts = {
              jsescOption: { minimal: true },
            };
          }

          return {
            babelOptions,
            includes,
            excludes,
          };
        };

        const { rootPath } = api.context;
        const { babelOptions, includes, excludes } = getBabelOptions(
          rootPath,
          config,
        );
        const useTsLoader = Boolean(config.tools.tsLoader);
        const rule = chain.module.rule(CHAIN_ID.RULE.JS);

        applyScriptCondition({
          rule,
          config,
          context: api.context,
          includes,
          excludes,
        });

        rule
          .test(useTsLoader ? JS_REGEX : mergeRegex(JS_REGEX, TS_REGEX))
          .use(CHAIN_ID.USE.BABEL)
          .loader(getCompiledPath('babel-loader'))
          .options(babelOptions);

        /**
         * If a script is imported with data URI, it can be compiled by babel too.
         * This is used by some frameworks to create virtual entry.
         * https://webpack.js.org/api/module-methods/#import
         * @example: import x from 'data:text/javascript,export default 1;';
         */
        if (config.source.compileJsDataURI) {
          chain.module
            .rule(CHAIN_ID.RULE.JS_DATA_URI)
            .mimetype({
              or: ['text/javascript', 'application/javascript'],
            })
            .use(CHAIN_ID.USE.BABEL)
            .loader(getCompiledPath('babel-loader'))
            .options(babelOptions);
        }

        addCoreJsEntry({ chain, config, isServer });
      },
    );
  },
});

/** Add core-js-entry to every entries. */
export function addCoreJsEntry({
  chain,
  config,
  isServer,
}: {
  chain: WebpackChain;
  config: NormalizedConfig;
  isServer: boolean;
}) {
  if (enableCoreJsEntry(config, isServer)) {
    const entryPoints = Object.keys(chain.entryPoints.entries() || {});
    const coreJsEntry = createVirtualModule('import "core-js";');

    for (const name of entryPoints) {
      chain.entry(name).prepend(coreJsEntry);
    }
  }
}
