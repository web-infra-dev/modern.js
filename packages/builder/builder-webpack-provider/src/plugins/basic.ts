import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of webpack
 */
export const builderPluginBasic = (): BuilderPlugin => ({
  name: 'builder-plugin-basic',

  setup(api) {
    api.modifyWebpackChain(async (chain, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      chain.context(api.context.rootPath);

      chain.mode(isProd ? 'production' : 'development');

      chain.merge({
        infrastructureLogging: {
          // Using `error` level to avoid `webpack.cache.PackFileCacheStrategy` logs
          level: 'error',
        },
      });

      // if the chunk size exceeds 1MiB, we will throw a warning
      chain.performance.maxAssetSize(1024 * 1024);
      chain.performance.maxEntrypointSize(1024 * 1024);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });
    });
  },
});
