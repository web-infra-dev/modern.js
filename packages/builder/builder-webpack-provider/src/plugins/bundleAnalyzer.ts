import { BuilderPlugin } from '../types';

export function builderPluginBundleAnalyzer(): BuilderPlugin {
  return {
    name: 'builder-plugin-bundle-analyzer',
    setup(api) {
      api.modifyWebpackChain(
        async (chain, { CHAIN_ID, getCompiledPath, target }) => {
          const config = api.getNormalizedConfig();
          // There are two ways to open the bundle analyzer:
          // 1. Set environment variable `BUNDLE_ANALYZE`
          // 2. Set performance.bundleAnalyze config
          if (
            !process.env.BUNDLE_ANALYZE &&
            !config.performance.bundleAnalyze
          ) {
            return;
          }
          const { default: BundleAnalyzer } = await import(
            getCompiledPath('webpack-bundle-analyzer')
          );
          chain
            .plugin(CHAIN_ID.PLUGIN.BUNDLE_ANALYZER)
            .use(BundleAnalyzer.BundleAnalyzerPlugin, [
              {
                analyzerMode: 'static',
                openAnalyzer: false,
                reportFilename: `report-${target}.html`,
                ...(config.performance.bundleAnalyze || {}),
              },
            ]);
        },
      );
    },
  };
}
