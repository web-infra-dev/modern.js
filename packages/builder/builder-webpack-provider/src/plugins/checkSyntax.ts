import {
  BuilderContext,
  BuilderTarget,
  DEFAULT_BROWSERSLIST,
  getBrowserslistWithDefault,
} from '@modern-js/builder-shared';
import type { BuilderPlugin, NormalizedConfig } from '../types';

export function builderPluginCheckSyntax(): BuilderPlugin {
  return {
    name: 'builder-plugin-check-syntax',

    setup(api) {
      api.modifyWebpackChain(async (chain, { isProd, target }) => {
        const config = api.getNormalizedConfig();
        const { checkSyntax } = config.security;

        if (
          !isProd ||
          ['node', 'web-worker'].includes(target) ||
          !checkSyntax
        ) {
          return;
        }

        const targets = await getCheckTargets(
          api.context,
          config,
          target,
          checkSyntax,
        );
        const { CheckSyntaxPlugin } = await import(
          '../webpackPlugins/CheckSyntaxPlugin'
        );
        chain.plugin(CheckSyntaxPlugin.name).use(CheckSyntaxPlugin, [
          {
            targets,
            exclude:
              typeof checkSyntax === 'object' ? checkSyntax.exclude : undefined,
          },
        ]);
      });
    },
  };
}

async function getCheckTargets(
  builderContext: BuilderContext,
  builderConfig: NormalizedConfig,
  builderTarget: BuilderTarget,
  checkSyntax: { targets: string[] } | true,
) {
  if (checkSyntax === true) {
    const browserslist = await getBrowserslistWithDefault(
      builderContext.rootPath,
      builderConfig,
      builderTarget,
    );

    return browserslist || DEFAULT_BROWSERSLIST[builderTarget];
  }
  return checkSyntax.targets;
}
