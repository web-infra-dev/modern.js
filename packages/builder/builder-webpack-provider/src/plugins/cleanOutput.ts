import type { BuilderPlugin } from '../types';

export const PluginCleanOutput = (): BuilderPlugin => ({
  name: 'builder-plugin-clean-output',

  setup(api) {
    const clean = async () => {
      const config = api.getNormalizedConfig();

      if (config.output.cleanDistPath) {
        const { emptyDir } = await import('@modern-js/utils');
        const { distPath } = api.context;
        await emptyDir(distPath);
      }
    };

    api.onBeforeBuild(clean);
    api.onBeforeStartDevServer(clean);
  },
});
