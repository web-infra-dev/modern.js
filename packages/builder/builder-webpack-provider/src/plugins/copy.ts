import type { CopyPluginOptions, BuilderPlugin } from '../types';

export const builderPluginCopy = (): BuilderPlugin => ({
  name: 'builder-plugin-copy',

  setup(api) {
    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
      const config = api.getNormalizedConfig();
      const { copy } = config.output;

      if (!copy) {
        return;
      }

      const { default: CopyPlugin } = await import(
        '../../compiled/copy-webpack-plugin'
      );

      const options: CopyPluginOptions = Array.isArray(copy)
        ? { patterns: copy }
        : copy;

      chain.plugin(CHAIN_ID.PLUGIN.COPY).use(CopyPlugin, [options]);
    });

    api.modifyWebpackConfig(async config => {
      const copyPlugin = config.plugins?.find(
        item => item.constructor.name === 'CopyPlugin',
      ) as unknown as CopyPluginOptions;

      if (copyPlugin) {
        const { fs } = await import('@modern-js/utils');

        // If the pattern.context directory not exists, we should remove CopyPlugin.
        // Otherwise the CopyPlugin will cause the webpack to re-compile.
        const isContextNotExists = copyPlugin.patterns.every(
          pattern =>
            typeof pattern !== 'string' &&
            pattern.context &&
            !fs.existsSync(pattern.context),
        );
        if (isContextNotExists) {
          config.plugins = config.plugins?.filter(
            item => item.constructor.name !== 'CopyPlugin',
          );
        }
      }
    });
  },
});
