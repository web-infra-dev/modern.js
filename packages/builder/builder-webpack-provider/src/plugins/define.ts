import _ from '@modern-js/utils/lodash';
import type { BuilderPlugin, SourceConfig } from '../types';

export const builderPluginDefine = (): BuilderPlugin => ({
  name: 'builder-plugin-define',

  async setup(api) {
    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
      const config = api.getNormalizedConfig();

      const builtinVars: NonNullable<SourceConfig['globalVars']> = {
        'process.env.NODE_ENV': process.env.NODE_ENV || 'development',
      };

      // Serialize global vars. User can customize value of `builtinVars`.
      const globalVars = { ...builtinVars, ...config.source.globalVars };
      const serializedVars = _.mapValues(globalVars, value =>
        JSON.stringify(value),
      );
      // Macro defines.
      const defineExprs = config.source.define;

      const { DefinePlugin } = await import('webpack');
      chain
        .plugin(CHAIN_ID.PLUGIN.DEFINE)
        .use(DefinePlugin, [{ ...serializedVars, ...defineExprs }]);
    });
  },
});
