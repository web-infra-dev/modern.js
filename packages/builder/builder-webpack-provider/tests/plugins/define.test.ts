import { describe, expect, it } from 'vitest';
import { createStubBuilder } from '@/stub';
import { builderPluginDefine } from '@/plugins/define';

describe('plugins/define', () => {
  it('should match snapshot', async () => {
    const globalVars = {
      'process.env.foo': 'foo',
      'import.meta.bar': {
        a: 'bar',
        b: false,
        c: { d: 42 },
      },
      'window.baz': [null, 'baz'],
    };
    const define = {
      'import.meta.test': false,
      cov: () => process.stdout.write('cov'),
    };
    const builder = await createStubBuilder({
      plugins: [builderPluginDefine()],
      builderConfig: { source: { globalVars, define } },
    });
    expect(await builder.unwrapWebpackConfig()).toMatchInlineSnapshot(`
      {
        "plugins": [
          DefinePlugin {
            "definitions": {
              "cov": [Function],
              "import.meta.bar": "{\\"a\\":\\"bar\\",\\"b\\":false,\\"c\\":{\\"d\\":42}}",
              "import.meta.test": false,
              "process.env.NODE_ENV": "\\"test\\"",
              "process.env.foo": "\\"foo\\"",
              "window.baz": "[null,\\"baz\\"]",
            },
          },
        ],
      }
    `);
  });
});
