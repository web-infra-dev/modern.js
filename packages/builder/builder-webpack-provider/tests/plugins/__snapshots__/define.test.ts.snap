// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/define > should allow globalVars to be a function 1`] = `
{
  "plugins": [
    DefinePlugin {
      "definitions": {
        "ENV": "\\"test\\"",
        "TARGET": "\\"web\\"",
        "process.env.ASSET_PREFIX": "\\"\\"",
        "process.env.NODE_ENV": "\\"test\\"",
      },
    },
  ],
}
`;
